### PR TITLE
ProjectM 3.1.12 (new formula)

### PR DESCRIPTION
**Pull request by revmischa**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

### CODE REVIEW

File: Formula/projectm.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+class Projectm < Formula
+  desc "Milkdrop-compatible music visualizer. Standalone app and Music.app plugin"
```
Comment (bayandin) [inline Formula/projectm.rb:2]: It looks like homebrew-casks fits better for this app.

https://docs.brew.sh/Acceptable-Formulae#stuff-that-builds-an-app

**Comment by bayandin** [comment]
> Getting an error on `brew test projectm`, not sure what the meaning is

Formula requires real test block instead of current stub

File: Formula/projectm.rb (hunk 1 above)
Comment (revmischa) [reply Formula/projectm.rb:2]: I think the description is inaccurate for the source build here, I'll update it. It builds a command-line app.

Screen Shot 2021-02-14 at 6 17 05 PM

**Comment by revmischa** [comment]
Not really sure what I'm doing wrong in my tests

**Comment by SMillerDev** [comment]
> Not really sure what I'm doing wrong in my tests

The test should be a command a user would issue. And check if the output is as expected. You're currently just checking if files exist (and they don't all exist).

**Comment by revmischa** [comment]
I don't understand why the exists check fails though

**Comment by SMillerDev** [comment]
> I don't understand why the exists check fails though

It doesn't really matter why, it's not a sufficient test either way so it'll have to be replaced.
We need a test that exercises the some of the functionality of the app. 

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

File: Formula/projectm.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  def install
+    system "./configure", "--disable-debug",
+                          "--disable-dependency-tracking",
+                          "--disable-silent-rules",
+                          "--prefix=#{prefix}"
+    system "make", "install"
+  end
+
+  test do
+    system bin/"projectM-unittest"
```
Comment (SMillerDev) [inline Formula/projectm.rb:20]: Is this something a user would regularly run?

Comment (revmischa) [reply Formula/projectm.rb:20]: It's not no

**Comment by revmischa** [comment]
Is this new test satisfactory?

**Comment by SMillerDev** [comment]
Can you squash all your commits into one with the same title as the PR?

**Comment by revmischa** [comment]
Squashed

**Review by SMillerDev — DISMISSED** [review]
Thanks @revmischa  ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉

**Review by carlocab — APPROVED** [review]

**Pull request closed without merging**